### PR TITLE
Add lenient mode to Version string parsing that doesn't require patch version

### PR DESCRIPTION
**Pull request by colincornaby**
Git repos are commonly leaving off the patch segments of version strings in tags. This practice is also commonly repeated in guides on Git tagging. This commit makes the patch version optional to improve compatibility with these Git repositories in SPM.

Because of the possible downstream effects, I've made the lenient version parsing optional, and it defaults to off. Right now it only allows the patch version to be left off.

I considered trying to keep changes entirely in SPM without touching the Version type. This seemed like the most maintainable and performant option.

### CODE REVIEW

File: Sources/TSCUtility/Version.swift
Comment (WowbaggersLiquidLunch) [inline Sources/TSCUtility/Version.swift:123]: If we're doing this, would it be better (or make sense) if the rules are made even more lenient to allow for just a single major version number? 

i.e.

```swift
guard requiredComponents.count == 3 || lenient else ...
```

Comment (colincornaby) [reply Sources/TSCUtility/Version.swift:123]: My original intention was to make repositories that mix two and three component version styles more compatible. Supporting a single version component would make SPM support repositories that don't technically version semantically, which is a larger change. But that behavior can still map to semantic versions, and it would maximize compatibility with existing repos.

**Comment by tomerd** [comment]
@colincornaby thanks for this proposal. do you think package manifest should also allow `x.0` to be parsed as `x.0.0`? If so, we would also need a change in SwiftPM itself: https://github.com/apple/swift-package-manager/blob/main/Sources/PackageDescription/Version.swift

File: Sources/TSCUtility/Version.swift
Comment (tomerd) [reply Sources/TSCUtility/Version.swift:123]: I tend to agree that from a conversion point of, tags are either 3 `x.0.0` or 2 `x.0` segment long. iow a single digit version tag would be pretty exceptional.

File: Tests/TSCUtilityTests/VersionTests.swift
Comment (tomerd) [inline Tests/TSCUtilityTests/VersionTests.swift:290]: format

**Comment by colincornaby** [comment]
@tomerd - I had noticed that there is a package manifest facing aspect to this. I'll be honest, this is my first pull request to Swift, so I was unsure how large of changes the project is interested in through pull request. I also didn't know what existing opinions project opinions were on this.

Short answer is: No, but I'd be willing to make those changes if that's desired by others. My motivation was that we had started to implement SPM manifests in some of our C++ projects. But those teams hadn't always been adding the patch version where it was 0 to their tags (e.g. they used 3.4, 3.4.1, 3.4.2, 3.5). I wanted to be able to introduce SPM manifests into repos without retraining teams. A quick survey found a lot of Git repos also dropping the bug fix version when it was 0.

I do also have changes to the SPM repo that would integrate these changes into Git tagging, but they're waiting on these changes. If it's desired for the manifest to also support two digits I can propose those changes in a pull request as well.

**Comment by tomerd** [comment]
thanks @colincornaby makes sense. I think the proposed change is valuable at the git tag parsing level. @neonichu wdyt?

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by tomerd** [comment]
@colincornaby CI is failing on something that looks like an issue with the tests?

```
/home/buildnode/jenkins/workspace/pr-swift-tools-support-core-linux/branch-main/swiftpm/Sources/SPMTestSupport/MockPackageContainer.swift:95:63: error: type 'Version' has no member 'init(string:)'
17:14:37         self._versions = dependencies.keys.compactMap(Version.init(string:)).sorted()
17:14:37                                                       ~~~~~~~ ^~~~~~~~~~~~~
17:14:37 /home/buildnode/jenkins/workspace/pr-swift-tools-support-core-linux/branch-main/swiftpm/Sources/SPMTestSupport/MockWorkspace.swift:123:25: error: generic parameter 'U' could not be inferred
17:14:37                 let v = version.flatMap(Version.init(string:))
17:14:37                         ^
17:14:37 Swift.Optional:6:28: note: in call to function 'flatMap'
17:14:37     @inlinable public func flatMap<U>(_ transform: (Wrapped) throws -> U?) rethrows -> U?
17:14:37                            ^
17:14:37 /home/buildnode/jenkins/workspace/pr-swift-tools-support-core-linux/branch-main/swiftpm/Sources/SPMTestSupport/MockWorkspace.swift:123:49: error: type 'Version' has no member 'init(string:)'
17:14:37                 let v = version.flatMap(Version.init(string:))
17:14:37                                         ~~~~~~~ ^~~~~~~~~~~~~
```

File: Sources/TSCUtility/Version.swift
Comment (tomerd) [inline Sources/TSCUtility/Version.swift:113]: since there are some dependencies in SwiftPM on the concrete ctor signature, lets structure this as:

```
init?(string: String) {
 self.init(string: string, lenient: false)
}

init?(string: String, lenient: Bool) {
 ...
}
```

Comment (WowbaggersLiquidLunch) [reply Sources/TSCUtility/Version.swift:113]: Do we really need the the `lenient` parameter? If the goal is to treat all x.y as x.y.0, then I don't we can use strict parsing anywhere.

Comment (colincornaby) [reply Sources/TSCUtility/Version.swift:113]: I'm happy to remove the lenient option if no one needs it as a solution.

Comment (tomerd) [reply Sources/TSCUtility/Version.swift:113]: imo its good to have this option for now so we can enable/disable it from SwiftPM side based on which tools version we are targeting - as this change could break existing packages. @neonichu wdyt?

Comment (neonichu) [reply Sources/TSCUtility/Version.swift:113]: I agree that the flag is useful.

Comment (WowbaggersLiquidLunch) [reply Sources/TSCUtility/Version.swift:113]: > imo its good to have this option for now so we can enable/disable it from SwiftPM side based on which tools version we are targeting - as this change could break existing packages.

This makes sense. I keep forgetting that SwiftPM needs to support previous versions.

Comment (colincornaby) [reply Sources/TSCUtility/Version.swift]: Thanks for the feedback and pointing me to exactly what the test failure was. I've made the changes and run the in-module tests locally.

Comment (tomerd) [reply Sources/TSCUtility/Version.swift]: >  I keep forgetting that SwiftPM needs to support previous versions.

that's the hard part :D

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by colincornaby** [comment]
Looks like the Linux build failed, but it's a module cache issue.

```
<unknown>:0: error: module file '/home/buildnode/jenkins/workspace/pr-swift-tools-support-core-linux/branch-main/build/buildbot_incremental/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/swift-driver/module-cache/2YIFRU0PHJR5U/CDispatch-SLYT5KF7B5CN.pcm' is out of date and needs to be rebuilt: signature mismatch
```

**Comment by WowbaggersLiquidLunch** [comment]
@swift-ci please clean test

**Comment by neonichu** [comment]
@swift-ci please test

**Review by tomerd — APPROVED** [review]

**Comment by WowbaggersLiquidLunch** [comment]
Sorry for asking this, but I wonder if it's OK to merge #214 first (after it's ready) and then rebase this PR on it? 

These 2 PRs conflict with each other, so one of these has to be rebased after the other is merged. It seems that it's much easier to rebase this PR on #214 than the vice versa, because #214 changes a lot more lines and the logic. And since most of the parsing logic is changed in #214, if this PR is merged first, everything here would have to be reimplemented again in #214. If we end up deciding to merge #214 first, I can help with rebasing this PR.

**Comment by tomerd** [comment]
> Sorry for asking this, but I wonder if it's OK to merge #214 first (after it's ready) and then rebase this PR on it?

> These 2 PRs conflict with each other, so one of these has to be rebased after the other is merged. It seems that it's much easier to rebase this PR on #214 than the vice versa, because #214 changes a lot more lines and the logic. And since most of the parsing logic is changed in #214, if this PR is merged first, everything here would have to be reimplemented again in #214. If we end up deciding to merge #214 first, I can help with rebasing this PR


sgtm, assuming we can get #214 over the finish line soon

**Comment by abertelrud** [comment]
> > Sorry for asking this, but I wonder if it's OK to merge #214 first (after it's ready) and then rebase this PR on it?
> 
> > These 2 PRs conflict with each other, so one of these has to be rebased after the other is merged. It seems that it's much easier to rebase this PR on #214 than the vice versa, because #214 changes a lot more lines and the logic. And since most of the parsing logic is changed in #214, if this PR is merged first, everything here would have to be reimplemented again in #214. If we end up deciding to merge #214 first, I can help with rebasing this PR
> 
> sgtm, assuming we can get #214 over the finish line soon

I think #214 looks good to me at this point, so can be merged if @neonichu agrees.

**Comment by colincornaby** [comment]
Following along - I'm fine also doing any cleanup work necessary after #214 gets merged in, or assisting others in that. I have changes in SPM waiting on this pull request to be merged, but no huge rush there.

**Comment by tomerd** [comment]
@colincornaby #214 is now merged, please rebase. thanks for your patience on this

**Comment by WowbaggersLiquidLunch** [comment]
@colincornaby Can I have access to this branch, to help with the rebase?

**Comment by colincornaby** [comment]
@WowbaggersLiquidLunch - I sent a collaboration invite over for my fork. I had started on the rebase, but in since you know the changes better I'll let you make the changes!

**Comment by WowbaggersLiquidLunch** [comment]
Just force-pushed 19d1c37eb8b239429f9eec85ce847e084f31b695.

I took the liberty of squashing the 3 previous commits on this branch, so that it's easier to rebase. Hopefully this is ok. I also saved a copy of the pre-rebase branch [here](https://github.com/WowbaggersLiquidLunch/swift-tools-support-core/tree/lenient-version-string-parsing-backup), just incase.

Changes as part of the rebase:

Instead of a new `Version.init?(string:lenient)`, I added the `usesLenientParsing` parameter to `Version.init(versionString:) throws`, because #214 moved all the parsing logic to the latter initializer and deprecated the former. I changed of parameter name to follow the [API guidelines](https://swift.org/documentation/api-design-guidelines/#boolean-assertions) more closely.

`usesLenientParsing` is also added to `VersionError.invalidVersionCoreIdentifiersCount` as an additional associated value, because the error message needs to know whether the minimum number of version core identifiers is 2 or 3.

Most of the additional lines of code are test cases.

File: Sources/TSCUtility/Version.swift
Comment (tomerd) [inline Sources/TSCUtility/Version.swift:105]: iirc we needed two constructors instead of one with default value because of downstream dependencies on the exact signatures (passed into map{})

Comment (WowbaggersLiquidLunch) [reply Sources/TSCUtility/Version.swift:105]: I think the one that downstream dependencies use is `init?(string:)`. This initializer still exists and calls `init?(_:)`, which then calls `init(versionString: versionString, usesLenientParsing: false) throws`.

Comment (colincornaby) [reply Sources/TSCUtility/Version.swift:105]: I have the change ready to commit, but also had the thought the downstream dependencies probably use the deprecated function. I could see the use in having a new version ready to cut over to, but that might get into project planning that I don't know the scope of.

**Comment by tomerd** [comment]
@swift-ci please test

File: Sources/TSCUtility/Version.swift
Comment (tomerd) [reply Sources/TSCUtility/Version.swift:105]: thanks @colincornaby and @WowbaggersLiquidLunch - lets get this merged first, then @colincornaby you can submit a PR to SwiftPM to use the new methods where appropriate.

Comment (colincornaby) [reply Sources/TSCUtility/Version.swift:105]: Sounds good! Thanks!

**Comment by tomerd** [comment]
thanks again thanks @colincornaby and @WowbaggersLiquidLunch

**Pull request merged**
